saturn_client/core.py

```diff
@@ -0,0 +1,161 @@
+"""Python library for interacting with the Saturn Cloud API.
+
+NOTE: This is an experimental library and will likely change in
+the future
+"""
+
+import json
+import logging
+
+from typing import Any, Dict, Optional
+from urllib.parse import urljoin
+
+import requests
+from .settings import Settings
+
+
+log = logging.getLogger("saturn-client")
+if log.level == logging.NOTSET:
+    log.setLevel(logging.INFO)
+
+
+class SaturnConnection:
+    """
+    Create a ``SaturnConnection`` to interact with the API.
+
+    :param url: URL for the SaturnCloud instance.
+    :param api_token: API token for authenticating the request to Saturn API.
+        Get from `/api/user/token`
+    """
+
+    _options = None
+    _saturn_api_version = None
+
+    def __init__(
+        self,
+        url: Optional[str] = None,
+        api_token: Optional[str] = None,
+    ):
+        """
+        Create a ``SaturnConnection`` to interact with the API.
+
+        :param url: URL for the SaturnCloud instance.
+        :param api_token: API token for authenticating the request to Saturn API.
+            Get from `/api/user/token`
+        """
+        self.settings = Settings(url, api_token)
+
+        # test connection to raise errors early
+        self._saturn_api_version = self.saturn_api_version
+
+    @property
+    def url(self):
+        """URL of Saturn instance"""
+        return self.settings.url
+
+    @property
+    def saturn_api_version(self):
+        """Version of Saturn API"""
```

Comment:
I'm not sold on calling this `api_version` when it's really "current release of Saturn".

The difference is that API versioning is a really specific, meaningful term and it's usually used in this context to indicate breaking changes, and usually users can choose when to upgrade from one to the other. So, for example, if we had `/api/v1/deployments`, requests to that endpoint would continue to work even after `/api/v2/deployments` was released, until it was eventually shut down, and your code could choose which one to hit.

If I saw the phrase `api_version` in a client library like this, I'd expect that that is what it's talking about.

This isn't that though. This is the current release of Saturn, which might or might not include breaking changes. And right now you can only have a single value of this property for any given `url`.

So at a minimum, I recommend making this `__saturn_api_version` so no customer code accesses it. But I also recommend changing the name to `__saturn_release` or something like that.

Reply:
Yeah I was thinking about that too. I wanted to distinguish it from the version of this libarry, so that's why I added in api. I think just calling it `_saturn_version` and making it private like you suggest is a good idea.